internal/component/database_observability/mysql/collector/explain_plan.go

```diff
@@ -0,0 +1,750 @@
+	dbConnection     *sql.DB
+	instanceKey      string
+	dbVersion        string
+	scrapeInterval   time.Duration
+	queryCache       []queryInfo
+	perScrapeRatio   float64
+	currentBatchSize int
+	entryHandler     loki.EntryHandler
+
+	logger  log.Logger
+	running *atomic.Bool
+	ctx     context.Context
+	cancel  context.CancelFunc
+}
+
+func NewExplainPlan(args ExplainPlanArguments) (*ExplainPlan, error) {
+	rs := args.DB.QueryRowContext(context.Background(), selectDbSchemaVersion)
+	if rs.Err() != nil {
+		return nil, rs.Err()
+	}
+
+	var dbVersion string
+	if err := rs.Scan(&dbVersion); err != nil {
+		return nil, err
+	}
+
+	return &ExplainPlan{
+		dbConnection:   args.DB,
+		instanceKey:    args.InstanceKey,
+		dbVersion:      dbVersion,
+		scrapeInterval: args.ScrapeInterval,
+		queryCache:     make([]queryInfo, 0),
+		perScrapeRatio: args.PerScrapeRatio,
+		entryHandler:   args.EntryHandler,
+		logger:         log.With(args.Logger, "collector", ExplainPlanName),
+		running:        atomic.NewBool(false),
+	}, nil
+}
+
+func (c *ExplainPlan) Name() string {
+	return ExplainPlanName
+}
+
+func (c *ExplainPlan) Start(ctx context.Context) error {
+	level.Info(c.logger).Log("msg", "collector started")
+
+	c.running.Store(true)
+	ctx, cancel := context.WithCancel(ctx)
+	c.ctx = ctx
+	c.cancel = cancel
+
+	go func() {
+		defer func() {
+			c.Stop()
+			c.running.Store(false)
+		}()
+
+		ticker := time.NewTicker(c.scrapeInterval)
+
+		for {
+			if err := c.fetchExplainPlans(c.ctx); err != nil {
+				level.Error(c.logger).Log("msg", "collector error", "err", err)
+			}
+
+			select {
+			case <-c.ctx.Done():
+				return
+			case <-ticker.C:
+				// continue loop
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (c *ExplainPlan) Stopped() bool {
+	return !c.running.Load()
+}
+
+func (c *ExplainPlan) Stop() {
+	c.cancel()
+}
+
+func (c *ExplainPlan) populateQueryCache(ctx context.Context) error {
+	rs, err := c.dbConnection.QueryContext(ctx, selectDigestsForExplainPlan)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to fetch digests for explain plans", "err", err)
+		return err
+	}
+	defer rs.Close()
+
+	// Populate cache
+	for rs.Next() {
+		if err := rs.Err(); err != nil {
+			level.Error(c.logger).Log("msg", "failed to iterate rs digests for explain plans", "err", err)
+			return err
+		}
+
+		var qi queryInfo
+		if err = rs.Scan(&qi.schemaName, &qi.digest, &qi.queryText); err != nil {
+			level.Error(c.logger).Log("msg", "failed to scan digest for explain plans", "err", err)
+			return err
+		}
+		c.queryCache = append(c.queryCache, qi)
+	}
+	// Calculate batch size based on current cache size
+	c.currentBatchSize = int(math.Ceil(float64(len(c.queryCache)) * c.perScrapeRatio))
+	level.Info(c.logger).Log("msg", "fetched digests", "count", len(c.queryCache), "batch_size", c.currentBatchSize)
+	return nil
+}
+
+func (c *ExplainPlan) fetchExplainPlans(ctx context.Context) error {
+	// If cache is empty, fetch all available queries
```

Comment:
Some of the comments seem to be attached to code that is quite self describing. Could you take out all but the ones for cases where you think the code is not self describing?

Reply:
Yeah, fair.. Pretty sure those are vestigial from when I was first writing this, and left them as notes to myself on what to implement 😂